### PR TITLE
Add triangulate Polygon Soup

### PR DESCRIPTION
**Pull request by MaelRL**
## Summary of Changes

- Introduce `CGAL::Polygon_mesh_processing::triangulate_polygons()`
- Remove CDT2 branch from `CGAL::Polygon_mesh_processing::triangulate_face[s]()`: hole filling now does the CDT2 check itself using a better check (PCA)
- Properly forward and document named parameters to `PMP::triangulate_hole_polylines()`
- Document non-degeneracy as a precondition of these algorithms
- Add some try-catching in CDT2 constraint insertions

## Release Management

* Affected package(s): `Polygon_mesh_processing`
* Issue(s) solved (if any): -
* Feature/Small Feature (if any): [Triangulate Polygons](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Triangulate_Polygons), pre-approved by [sloriot](https://cgal.geometryfactory.com/CGAL/Members/wiki/User:Sloriot) -- 2023/03/31
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership: no change

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h [hunk 1]
```diff
@@ -63,366 +53,187 @@ struct Default_visitor {
...
+      Point_ref p0 = get(vpm, target(v0, pmesh));
+      v1 = next(v0, pmesh);
+      Point_ref p1 = get(vpm, target(v1, pmesh));
+      v2 = next(v1, pmesh);
+      Point_ref p2 = get(vpm, target(v2, pmesh));
+      v3 = next(v2, pmesh);
+      Point_ref p3 = get(vpm, target(v3, pmesh));
+
+      /* Chooses the diagonal that will split the quad in two triangles that maximize
+       * the scalar product of of the un-normalized normals of the two triangles.
```
Comment (albert-github) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h:209]: of of -> of

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h [hunk 2]
```diff
@@ -541,41 +332,40 @@ bool triangulate_face(typename boost::graph_traits<PolygonMesh>::face_descriptor
...
-                                  get_parameter(np, internal_np::visitor),
-                                  Triangulate_faces::Default_visitor<PolygonMesh>());
-
-  internal::Triangulate_modifier<PolygonMesh, VPMap, Kernel, Visitor> modifier(vpmap, traits);
-  return modifier(face_range, pmesh, use_cdt, visitor);
+  using face_descriptor = typename boost::graph_traits<PolygonMesh>::face_descriptor;
+
+  bool result = true;
+
+  // One need to store facet handles into a vector, because the list of
```
Comment (albert-github) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h:356]: need -> needs
facet -> the facet

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h [hunk 3]
```diff
@@ -626,10 +422,255 @@ bool triangulate_faces(PolygonMesh& pmesh,
...
+    const std::size_t original_size = polygon.size();
+    if(original_size == 4)
+    {
+      Point_ref p0 = get(pm, points[polygon[0]]);
+      Point_ref p1 = get(pm, points[polygon[1]]);
+      Point_ref p2 = get(pm, points[polygon[2]]);
+      Point_ref p3 = get(pm, points[polygon[3]]);
+
+      /* Chooses the diagonal that will split the quad in two triangles that maximize
+       * the scalar product of of the un-normalized normals of the two triangles.
```
Comment (albert-github) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h:556]: of of -> of

(maybe also on other places)

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h [hunk 4]
```diff
@@ -626,10 +422,255 @@ bool triangulate_faces(PolygonMesh& pmesh,
...
+} // namespace internal
+
+/**
+* \ingroup PMP_meshing_grp
+*
+* triangulates all polygons of a polygon soup. This function depends on the package \ref PkgTriangulation2.
+*
+* @tparam PointRange a model of `ConstRange`. The value type of its iterator is the point type.
+* @tparam PolygonRange a model of the concept `SequenceContainer` and `Swappable`,
+*                      whose `value_type` is itself a model of the concepts `SequenceContainer`
```
Comment (albert-github) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h:600]: concepts -> concept

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h [hunk 5]
```diff
@@ -626,10 +422,255 @@ bool triangulate_faces(PolygonMesh& pmesh,
...
+
+} // namespace internal
+
+/**
+* \ingroup PMP_meshing_grp
+*
+* triangulates all polygons of a polygon soup. This function depends on the package \ref PkgTriangulation2.
+*
+* @tparam PointRange a model of `ConstRange`. The value type of its iterator is the point type.
+* @tparam PolygonRange a model of the concept `SequenceContainer` and `Swappable`,
```
Comment (albert-github) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h:599]: concept -> concepts

**Review by sloriot — APPROVED** [review]

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.6-Ic-216

**Pull request merged**